File: app/src/main/java/com/iceteaviet/fastfoodfinder/ui/main/map/NearByStoreAdapter.java
package com.iceteaviet.fastfoodfinder.ui.main.map;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.google.android.gms.maps.model.LatLng;
import com.iceteaviet.fastfoodfinder.R;
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store;
import com.iceteaviet.fastfoodfinder.utils.LocationUtils;
import com.iceteaviet.fastfoodfinder.utils.ui.UiUtils;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by Genius Doan on 11/9/2016.
 */
public class NearByStoreAdapter extends RecyclerView.Adapter<NearByStoreAdapter.StoreViewHolder> {
    private LatLng currCameraPosition;
    private List<Store> mListStore;
    private StoreListListener listener;

    public NearByStoreAdapter() {
        mListStore = new ArrayList<>();
    }

    public void setOnStoreListListener(StoreListListener listener) {
        this.listener = listener;
    }

    public void setCurrCameraPosition(LatLng currCameraPosition) {
        this.currCameraPosition = currCameraPosition;
    }

    public void addStore(Store store) {
        mListStore.add(store);
        notifyItemInserted(mListStore.size() - 1);
    }

    public void addStores(List<Store> listStores) {
        int position = mListStore.size();
        mListStore.addAll(listStores);
        notifyItemRangeInserted(position, mListStore.size());
    }

    public void setStores(List<Store> listStores) {
        mListStore.clear();
        mListStore.addAll(listStores);
        notifyDataSetChanged();
    }

    @Override
    public StoreViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View convertView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_store, parent, false);
        return new StoreViewHolder(convertView, listener, mListStore);
    }

    @Override
    public void onBindViewHolder(StoreViewHolder holder, int position) {
        Store store = mListStore.get(position);
        holder.setData(store, LocationUtils.calcDistance(currCameraPosition, store.getPosition()));
    }

    @Override
    public int getItemCount() {
        return mListStore.size();
    }

    public interface StoreListListener {
        void onItemClick(Store store);
    }

    static class StoreViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.iv_item_store)
        ImageView logo;
        @BindView(R.id.tv_item_storename)
        TextView storeName;
        @BindView(R.id.tv_item_address)
        TextView storeAddress;
        @BindView(R.id.tv_item_distance)
        TextView storeDistance;


        public StoreViewHolder(View itemView, final StoreListListener listener, final List<Store> storeList) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onItemClick(storeList.get(getAdapterPosition()));
                }
            });
        }

        public void setData(Store store, double distance) {
            Glide.with(logo.getContext())
                    .load(UiUtils.getStoreLogoDrawableId(store.getType()))
                    .into(logo);
            storeName.setText(store.getTitle());
            storeAddress.setText(store.getAddress());
            storeDistance.setText(new DecimalFormat("##.## Km").format(distance));
        }
    }
}

File: app/src/main/kotlin/com/iceteaviet/fastfoodfinder/ui/main/map/NearByStoreAdapter.kt
package com.iceteaviet.fastfoodfinder.ui.main.map

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.android.gms.maps.model.LatLng
import com.iceteaviet.fastfoodfinder.R
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store
import com.iceteaviet.fastfoodfinder.ui.main.StoreDiffCallback
import com.iceteaviet.fastfoodfinder.utils.calcDistance
import com.iceteaviet.fastfoodfinder.utils.formatDistance
import com.iceteaviet.fastfoodfinder.utils.ui.getStoreLogoDrawableId
import kotlinx.android.synthetic.main.item_store.view.*
import java.util.*

/**
 * Created by Genius Doan on 11/9/2016.
 *
 * @see androidx.recyclerview.widget.ListAdapter
 */

@Deprecated("due to the publish of ListAdapter\n")
class NearByStoreAdapter : RecyclerView.Adapter<NearByStoreAdapter.StoreViewHolder>() {
    private var currCameraPosition: LatLng? = null
    private val mListStore: MutableList<Store>
    private var listener: StoreListListener? = null

    init {
        mListStore = ArrayList()
    }

    fun setOnStoreListListener(listener: StoreListListener) {
        this.listener = listener
    }

    fun setCurrCameraPosition(currCameraPosition: LatLng) {
        this.currCameraPosition = currCameraPosition
    }

    fun addStore(store: Store) {
        mListStore.add(store)
        notifyItemInserted(mListStore.size - 1)
    }

    fun addStores(listStores: List<Store>) {
        val position = mListStore.size
        mListStore.addAll(listStores)
        notifyItemRangeInserted(position, mListStore.size)
    }

    fun setStores(listStores: List<Store>) {
        val diffResult = DiffUtil.calculateDiff(StoreDiffCallback(this.mListStore, listStores))

        mListStore.clear()
        mListStore.addAll(listStores)

        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(@NonNull parent: ViewGroup, viewType: Int): StoreViewHolder {
        val convertView = LayoutInflater.from(parent.context).inflate(R.layout.item_store, parent, false)
        return StoreViewHolder(convertView, listener, mListStore)
    }

    override fun onBindViewHolder(@NonNull holder: StoreViewHolder, position: Int) {
        val store = mListStore[position]
        holder.setData(store, calcDistance(currCameraPosition!!, store.getPosition()))
    }

    override fun getItemCount(): Int {
        return mListStore.size
    }

    interface StoreListListener {
        fun onItemClick(store: Store)
    }

    class StoreViewHolder(itemView: View, listener: StoreListListener?, storeList: List<Store>) : RecyclerView.ViewHolder(itemView) {
        var logo: ImageView = itemView.iv_item_store
        var storeName: TextView = itemView.tv_item_storename
        var storeAddress: TextView = itemView.tv_item_address
        var storeDistance: TextView = itemView.tv_item_distance


        init {
            itemView.setOnClickListener {
                listener?.onItemClick(storeList[adapterPosition])
            }
        }

        fun setData(store: Store, distance: Double) {
            Glide.with(logo.context)
                    .load(getStoreLogoDrawableId(store.type))
                    .into(logo)
            storeName.text = store.title
            storeAddress.text = store.address
            storeDistance.text = formatDistance(distance)
        }
    }
}
